client/linuxabi/linux_abi.go

```diff
@@ -106,87 +112,185 @@ type SnpReportReq struct {
 	FwErr uint64
 }
 
+// ABI returns an object that can cross the ABI boundary and copy back changes to the original
+// object.
+func (r *SnpUserGuestRequest) ABI() BinaryConversion {
+	result := &snpUserGuestRequestConversion{
+		reqConv:  r.ReqData.ABI(),
+		respConv: r.RespData.ABI(),
+	}
+	result.abi.GuestMsgVersion = guestMsgVersion
+	result.abi.ReqData = result.reqConv.Pointer()
+	result.abi.RespData = result.respConv.Pointer()
+	return result
+}
+
+// Pointer returns a pointer to the object that crosses the ABI boundary.
+func (r *snpUserGuestRequestConversion) Pointer() unsafe.Pointer {
+	return unsafe.Pointer(&r.abi)
+}
+
+// Finish writes back the FwErr and any changes to the request or response objects.
+func (r *snpUserGuestRequestConversion) Finish(b BinaryConvertible) error {
+	s, ok := b.(*SnpUserGuestRequest)
+	if !ok {
+		return fmt.Errorf("Finish argument is %v. Expects a *SnpUserGuestRequestSafe", reflect.TypeOf(b))
+	}
+	if err := r.reqConv.Finish(s.ReqData); err != nil {
+		return fmt.Errorf("could not finalize request data: %v", err)
+	}
+	if err := r.respConv.Finish(s.RespData); err != nil {
+		return fmt.Errorf("could not finalize response data: %v", err)
+	}
+	s.FwErr = r.abi.FwErr
+	return nil
+}
+
+// BinaryConversion is an interface that abstracts a "stand-in" object that passes through an ABI
+// boundary and can finalize changes to the original object.
+type BinaryConversion interface {
+	Pointer() unsafe.Pointer
+	Finish(BinaryConvertible) error
+}
+
+// BinaryConvertible is an interface for an object that can produce a partner BinaryConversion
+// object to allow its representation to pass the ABI boundary.
+type BinaryConvertible interface {
+	ABI() BinaryConversion
+}
+
 // Ioctl performs the ioctl Linux syscall with the sev-guest Linux ABI unsafe pointer
 // manipulation contained all in this call.
-func Ioctl(fd int, command uintptr, sreq *SnpUserGuestRequestSafe) (uintptr, error) {
-	// Limit unsafe pointers to this scope by converting internal types to ABI types before a
-	// raw ioctl call, and converting back.
-	safeReqData := sreq.ReqData
-	var reqData interface{}
-	switch extReq := safeReqData.(type) {
-	case *SnpExtendedReportReqSafe:
-		var certsAddress uint64
-		if len(extReq.Certs) > 0 {
-			certsAddress = uint64(uintptr(unsafe.Pointer(&extReq.Certs[0])))
-		}
-		reqData = &SnpExtendedReportReq{
-			Data:         extReq.Data,
-			CertsAddress: certsAddress,
-			CertsLength:  extReq.CertsLength,
-		}
-	}
-	abi := SnpUserGuestRequest{
-		ReqData:  uint64(uintptr(unsafe.Pointer(&reqData))),
-		RespData: uint64(uintptr(unsafe.Pointer(&sreq.RespData))),
+func Ioctl(fd int, command uintptr, req *SnpUserGuestRequest) (uintptr, error) {
+	abi := req.ABI()
+	result, _, errno := unix.Syscall(unix.SYS_IOCTL, uintptr(fd), command, uintptr(abi.Pointer()))
+	abi.Finish(req)
+
+	// Work around a kernel bug with writing uninitialized memory back on non-EIO.
```

Comment:
Should we file a bug (or maybe a github issue?) explaining this issue in a bit more detail so we can fix it up eventually?

Also, it seems a bit concerning that we're checking that error is NOT EIO and then overwriting the fwerr. This seems like it could be masking a lot of errors in theory.

Reply:
I've reported it upstream on the commit that added the code. How else would you like this handled?

Reply:
Ack. So the internally, we'd:
1. File a bug.
2. Link the bug here w/ a TODO.


I'm trying to figure out the right precedent to set externally. For example, if we receive pull requests form outside of Google. I think we should:

1. Open a github issue. In this case, the github issue would link the upstream feedback that you've sent.
2. Add a comment w/ a TODO linking the github issue.

WDYT?

Reply:
Done.